src/Week4/HW04.hs

```diff
@@ -0,0 +1,103 @@
+{-# OPTIONS_GHC -Wall #-}
+module Week4.HW04 where
+
+import Data.List
+
+newtype Poly a = P [a]
+
+-- Exercise 1 -----------------------------------------
+
+x :: Num a => Poly a
+x = P [0, 1]
+
+-- Exercise 2 ----------------------------------------
+
+-- reverse input lists, drop all leading 0s. check for equality.
+
+
+instance (Num a, Eq a) => Eq (Poly a) where
+    (==) (P a) (P b) = helper a ==  helper b
+        where 
+            helper :: (Eq a, Num a) => [a] -> [a]
+            helper = dropWhile (==0) . reverse 
+ 
+-- Exercise 3 -----------------------------------------
+
+-- P [1,2,3] == [3x^2 + 2x + 1]
+
+instance (Num a, Eq a, Show a) => Show (Poly a) where
+    show (P []) = ""
+    show (P b) = plusLst
+        where 
+            helper :: (Num a, Eq a, Show a) => Integer -> a -> String
+            helper _ n  | n == 0 = ""
+            helper 0 coeff = show coeff
+            helper 1 1 = "x"
+            helper 1 (-1) = "-x"
+            helper 1 coeff = show coeff ++ "x"
+            helper degree 1 = "x^" ++ show degree
+            helper degree coeff = show coeff ++ "x^" ++ show degree
```

Comment:
There should probably be special case for coeff -1 here as well. How will this show e.g. `P [0,0,-1]`?